keps/sig-network/2438-dual-stack-apiserver/README.md

```diff
@@ -0,0 +1,521 @@
+Service to dual-stack, but then the second master reboots before
+upgrading, causing the old apiserver to restart and revert the Service
+definition back to its single-stack form, before then being upgraded
+and changing it back to dual-stack again.
+
+However, at all times, the Service should fully usable; it should
+never claim to be dual-stack unless there actually are dual-stack
+apiservers running.
+
+(The one way that something could go wrong would be if a spurious
+apiserver were to start up after the upgrade/downgrade completed. Eg,
+after upgrading all apiservers to a dual-stack release, if an old
+single-stack-only apiserver were to start up, but then exit, it might
+spuriously change the Service back to single-stack before exiting.
+It's not clear why this would happen, and at any rate, it would be
+fixed as soon as any of the real apiservers restarted again.)
+
+#### Handling of "`kubernetes`" Endpoints and EndpointSlices
+
+For the "`kubernetes`" `Endpoints` (_not_ `EndpointSlice`), there
+should be no flapping; both old and new apiservers will agree at all
+times about what endpoints are available _in the cluster's primary IP
+family_, and thus will always agree on the contents of the
+single-stack `Endpoints` resource.
+
+For `EndpointSlice`, current versions of kube-apiserver create a
+single `EndpointSlice` object named "`kubernetes`", and assume that
+other apiservers will do the same; they don't look at any other
+`EndpointSlice` objects when reconciling. To create dual-stack
+EndpointSlices, a new apiserver would need to create two objects with
+different names. Even if new apiservers use the same name as the old
+apiserver does for the primary-IP-family slice, old apiservers would
+never see the slice it creates for the secondary IP family. This means
+there would be no flapping, but also means that a stale
+secondary-IP-family `EndpointSlice` could be left behind after a
+downgrade if the new apiserver exited uncleanly.
+
+This could be fixed by making the apiserver EndpointSlice reconciler
+smarter about what slices it reconciles ([kubernetes #101070]).
+Assuming this happened 2 releases before dual-stack apiserver support
+was added, then there would be no problems on downgrade, though there
+would be a problem with the old and new apiservers fighting over the
+existence of the secondary-IP-family `EndpointSlice` during the
+upgrade/downgrade process. This would eventually be resolved once all
+of the apiservers were running the same version, and the intermediate
+states would always be usable (ie, there might sometimes be actual
+IPv6 endpoints that were un-advertised, but there would never be
+advertised IPv6 endpoints that didn't work).
+
+Failing that, the release notes would have to document that in the
+event of a downgrade it might be necessary to manually clean the
+EndpointSlices up.
+
+[kubernetes #101070]: https://github.com/kubernetes/kubernetes/issues/101070
+
+### Version Skew Strategy
+
```

Comment:
I think that this sections refer to run several apiservers in parallel with different versions, that you touch in the upgrade/downgrade section

Reply:
No, the template asks "how will the component handle version skew with other components?" As you said, self-version-skew is covered in the upgrade/downgrade section.